### PR TITLE
feat(AppShell): add runtime parser

### PR DESCRIPTION
**Pull request by mgechev**
The Angular DI is dropped in this implementation for achieving minimal bundle size.

Code example:

```js
var ngShellParser;

const SHELL_PARSER_CACHE_NAME = 'bar';
const APP_SHELL_URL = 'shell.html';
const NO_RENDER_CSS_SELECTOR = '[no-render]';
const ROUTE_DEFINITIONS = [
      '/bar.html',
      '/index.html',
      '/profile/:id'
    ];

self.addEventListener('install', function (event) {
  const parser = System.import('app-shell/index')
        .then(module => {
          ngShellParser = module.shellParserFactory({
            SHELL_PARSER_CACHE_NAME,
            APP_SHELL_URL,
            NO_RENDER_CSS_SELECTOR,
            ROUTE_DEFINITIONS
          });
          return ngShellParser;
        })
        .then(() => ngShellParser.fetchDoc())
        .then(res => ngShellParser.parseDoc(res))
        .then(strippedResponse => {
          return caches.open(SHELL_PARSER_CACHE_NAME)
            .then(cache => {
              return cache.put(APP_SHELL_URL, strippedResponse);
            });
        })
        .catch(e => {
          console.error(e);
        });
  event.waitUntil(parser);
});

self.addEventListener('fetch', function(event) {
  event.respondWith(
    ngShellParser.match(event.request)
      .then(response => {
        if (response) return response;
        return caches.match(event.request)
          .then(function(response) {
            if (response) {
              return response;
            }
            return fetch(event.request);
          })
      })
  );
});
```

### CODE REVIEW

File: app-shell/src/app/shell-parser/shell-parser.spec.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+  <content class="shell-no-render">
+    <p>Hello world</p>
+  </content>
+  <div class="shell-no-render bar baz">
+  </div>
+</body>
+</html>
+`;
+
+const expectedResponse1_1 = `
```
Comment (jeffbcross) [inline app-shell/src/app/shell-parser/shell-parser.spec.ts]: Could you give these more descriptive names?

File: app-shell/src/app/shell-parser/shell-parser.spec.ts [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+      const parser = createMockedWorker(mockScope);
+      mockScope.mockResponses[SHELL_PARSER_DEFAULT_CONFIG.APP_SHELL_URL] = new MockResponse('foo');
+      parser.fetchDoc()
+        .then((res: MockResponse) => res.text())
+        .then((data: string) => {
+          expect(data).toBe('foo');
+          done();
+        });
+    });
+    it('should use the configured url when set', (done: any) => {
```
Comment (jeffbcross) [inline app-shell/src/app/shell-parser/shell-parser.spec.ts]: Could you add a line break between all specs?

**Comment by jeffbcross** [comment]
I reviewed it as well as I could, with the limitations of how much information I can keep in my mind at one time. I added a couple of nits. You can go ahead and merge once those comments are taken care of.

Some follow-ups after this PR:
 * Add e2e tests using the tooling @alxhub is adding in #61 
 * Case-sensitive-capable parser
 * Add a quick recipe showing how to use this, inside a new folder at the root of the project, `mobile-toolkit/cookbook/parse-app-shell-from-universal.md`. Eventually, we'll have a cookbook on mobile.angular.io based on recipes in this folder.

**Comment by mgechev** [comment]
Everything is fixed.

**Comment by mgechev** [comment]
I added some updates about the case sensitivity of the parser https://github.com/angular/mobile-toolkit/issues/60.

**Pull request merged**